py/torch_tensorrt/dynamo/conversion/impl/quantize.py

```diff
@@ -28,42 +28,73 @@ def quantize(
     """
 
     with unset_fake_temporarily():
-        if isinstance(input_tensor, TRTTensor) and input_tensor.dtype not in (
-            trt.float32,
-            trt.float16,
-        ):
-            raise ValueError(
-                f"quantize converter received an input of {input_tensor.dtype} type. Supported types: float32 | float16"
-            )
+        if isinstance(input_tensor, (torch.Tensor, TRTTensor)):
+            if input_tensor.dtype not in (
+                trt.float32,
+                trt.float16,
+                trt.bfloat16,
+                torch.bfloat16,
+                torch.float16,
+                torch.float32,
+            ):
+                raise ValueError(
+                    f"quantize converter received an input of {input_tensor.dtype} type. Supported types: float32 | float16 | bfloat16"
+                )
             if num_bits != 8 or exponent_bits not in (0, 4):
                 raise ValueError(
                     f"quantize converter currently only accept INT8 or FP8 based quantize, got {num_bits=}, {exponent_bits=}"
                 )
+        else:
+            raise ValueError(
+                f"quantize converter received an input of {type(input_tensor)} type. Supported types: torch.Tensor | TRTTensor"
+            )
+
         if num_bits == 8 and exponent_bits == 0:
+            dtype = trt.DataType.INT8
             max_bound = 127
         elif num_bits == 8 and exponent_bits == 4:
+            dtype = trt.DataType.FP8
             max_bound = 448
 
         amax = to_torch(amax, None)
+        axis = None
+        # int8 weight quantization is per-channel quantization(it can have one or multiple amax values)
+        if dtype == trt.DataType.INT8 and ".weight_quantizer" in name:
+            # if the amax has more than one element, calculate the axis, otherwise axis value will be ignored
+            if amax.numel() > 1:
+                amax_init_shape = amax.shape
+                amax = amax.squeeze().data
+                assert (
+                    len(amax.shape) == 1
+                ), f"TensorRT does not support multi-axis quantization. {name=} {amax_init_shape=} {amax.shape=} "
+                axis = list(amax_init_shape).index(list(amax.shape)[0])
+                assert (
+                    axis == 0
+                ), f"{name=} {amax=} is per-channel quantization, expected axis to be 0, but got {axis=}"
+        else:
+            # int8 activation and fp8 weight/activation quantization is per-tensor quantization, it can only have single amax value
+            assert (
+                amax.numel() == 1
+            ), f"{name=} is per-tensor quantization, expected amax is a singular value, but got {amax.shape=}"
         scale = torch.divide(amax, max_bound)
+        scale.masked_fill_(scale == 0, 1.0)
```

Comment:
what is this doing ? Also is there a scenario where amax can be an ITensor ?

Reply:
But if it is zero, we should error out right ? Otherwise the scale of 1.0 will be incorrect and would result in a silent accuracy issue ?